src/shogun/classifier/NearestCentroid.cpp

```diff
@@ -38,48 +38,59 @@ namespace shogun{
 		m_is_trained=false;
 	}
 
-
-	bool NearestCentroid::train_machine(std::shared_ptr<Features> data)
+	std::shared_ptr<Machine> NearestCentroid::fit(
+	    std::shared_ptr<Features> feat, std::shared_ptr<Labels> lab)
 	{
-		ASSERT(m_labels)
-		ASSERT(distance)
-		if (data)
-		{
-			if (m_labels->get_num_labels() != data->get_num_vectors())
-				error("Number of training vectors does not match number of labels");
-			distance->init(data, data);
-		}
-		else
-		{
-			data = distance->get_lhs();
-		}
+		// ASSERT(distance)
+		require(distance, "distance not set");
+		require(
+		    lab->get_num_labels() == feat->get_num_vectors(),
+		    "Number of training vectors does not match number of labels");
+		distance->init(feat, feat);
 
-		auto multiclass_labels = m_labels->as<MulticlassLabels>();
-		auto dense_data = data->as<DenseFeatures<float64_t>>();
+		auto multiclass_labels = lab->as<MulticlassLabels>();
+		auto dense_data = feat->as<DenseFeatures<float64_t>>();
 
 		int32_t num_classes = multiclass_labels->get_num_classes();
 		int32_t num_feats = dense_data->get_num_features();
 
 		SGMatrix<float64_t> centroids(num_feats, num_classes);
 		SGVector<int64_t> num_per_class(num_classes);
-
+		linalg::zero(centroids);
+		linalg::zero(num_per_class);
 		auto iter_labels = multiclass_labels->get_labels().begin();
+		std::map<int64_t, int64_t> label_to_index;
```

Comment:
why not unordered_map and then reserve num_classes?

Reply:
this labels to index, this is a problem that re-occurs in other places, so I'd not solve it here.
Actually, we have some mechanisms in place that convert labels internally (code is e.g. in multiclass labels).

Reply:
@karlnapf  didn't you start a PR that does this mapping?

Reply:
yes, there is something merged even.
But we need to re-think/redesign parts of it as it had issues ...